src/macros-by-example.adoc

```diff
@@ -0,0 +1,823 @@
+doesn't exist.  We can’t use null because `null` is a value!  In much the same way that Ion
+differentiates a null-field from an absent field (_e.g._, `{a:null}` versus `{}`), macros need to
+differentiate the absence of value from all other values in the Ion data model.
+
+We’ll use the term _void_ most accurately as a synonym for “empty stream”, but you might catch us
+talking as if void were a value, when it’s actually no value at all.  We’ll even mint the word
+_voidable_ to describe parameters that can accept zero values, like the ``*\...*``s above. This
+neologism will prove useful, you’ll see.
+
+The special-case E-expression `(:)` is synonymous with `(:void)`
+and exists solely as a more succinct argument expression for voidable parameters.
+
+
+=== Other Cardinalities
+
+For a language that’s thoroughly stream-oriented, rest-parameters don’t offer much flexibility
+with respect to cardinality.  The `*\...*` modifier can only be the final parameter, meaning that
+only that parameter can be invoked with more than one subexpression and produce more than one
+value.  So far other parameters must have exactly-one value, which is fairly limiting.
+
+In fact, the macro system provides four distinct cardinality constraints: exactly-one,
+zero-or-one, zero-or-more, and one-or-more.  They each have modifier symbols that can be used in
+signatures, namely `*!*`, `*?*`, `***`, and `*+*`. The former is the default, so the signatures
+`[arg]`, `[(*any* arg)]`, and `[(*any!* arg)]` are all equivalent.
+
+
+==== zero-or-one
+
+The zero-or-one cardinality, denoted in signatures by the modifier `*?*`, is much like the default
+exactly-one cardinality, except the parameter is voidable.  That is, it accepts an empty-stream
+argument as a way to denote an absent parameter.
+
+[{mrk}]
+----
+(*public* temperature
+  [(*decimal* degrees), (*symbol?* scale)]
+  {degrees: degrees, scale: scale})
+----
+
+Since the scale is voidable, we can pass it void:
+
+----
+(:temperature 96 F)    ⇒ {degrees:96, scale:F}
+(:temperature 283 (:)) ⇒ {degrees:283}
+----
+
+Note that the result’s `scale` field has disappeared because no value was provided.  It would be
+more useful to fill in a default value, and to do that we introduce another special form that can
+detect void:
+
+[{mrk}]
+----
+(*public* temperature
+  [(*decimal* degrees), (*symbol?* scale)]
+  {degrees: degrees, scale: (*if_void* scale (*literal* K) scale)})
+----
+----
+(:temperature 96 F)    ⇒ {degrees:96,  scale:F}
+(:temperature 283 (:)) ⇒ {degrees:283, scale:K}
+----
+
+The `*if_void*` form is if/then/else syntax testing stream emptiness. It has three sub-expressions,
+the first being a stream to check. If and only if that stream is void (_i.e._, it produces no
+values), the second sub-expression is expanded and its results are returned by the `*if_void*`
+expression. Otherwise, it produces at least one value, so the third sub-expression is expanded and
+returned.
+
```

Comment:
If I get my way, `(void (void))` will cause a compile time error when that macro definition is encountered, since the code is malformed in a statically-detectable way.   I'll take this as a request for a first-class way for macros to signal an error during their expansion.

But yeah, the compiler in my head thinks your code would work as intended.

Reply:
Maybe we need another special form. E.g.: `(err "At least one of 'a' and 'b' must be provided")`